src/components/DataTypes/date-time.tsx

```diff
@@ -30,64 +26,36 @@ interface DateTimeFilter extends JSONSchema {
 	};
 }
 
-export const decodeFilter = (
-	filter: DateTimeFilter,
-): {
-	field: string;
-	operator: OperatorSlug;
-	value: string;
-} | null => {
-	if (!filter.properties) {
-		return null;
-	}
-
-	const keys = Object.keys(filter.properties);
-	if (!keys.length) {
-		return null;
-	}
-
-	const field = keys[0];
-	const operator = filter.title;
-	let value: string;
-
-	if (operator === 'is') {
-		value = filter.properties[field].const!;
-	} else if (operator === 'is_before') {
-		value = filter.properties[field].formatMaximum!;
-	} else if (operator === 'is_after') {
-		value = filter.properties[field].formatMinimum!;
-	} else {
-		return null;
-	}
-
-	return {
-		field,
-		operator,
-		value,
-	};
-};
-
 export const createFilter = (
+	key: string,
 	field: string,
-	operator: OperatorSlug,
+	operator: Operator,
 	value: any,
 	schema: JSONSchema,
+	recursive: boolean = false,
+	refScheme?: string,
 ): DateTimeFilter => {
 	const { title } = schema;
-	const base: DateTimeFilter = {
-		$id: randomString(),
-		title: operator,
-		description: getJsonDescription(
-			title || field,
-			operators[operator].getLabel(schema),
-			value,
-		),
-		type: 'object',
-	};
+	const operatorSlug = operator.slug as OperatorSlug;
+	const base: DateTimeFilter = !recursive
+		? {
+				$id: randomString(),
+				title: operatorSlug,
+				description: getJsonDescription(
+					key,
+					title,
+					field,
+					operator,
+					value,
+					refScheme,
+				),
+				type: 'object',
+		  }
+		: ({} as DateTimeFilter);
 
 	const normalizedValue = normalizeDateTime(value);
 
-	if (operator === 'is') {
+	if (operatorSlug === 'is') {
```

Comment:
Looks like we can just do the following, no?:
```
switch(operatorSlug) {
  case 'is':
  case 'is_before':
  case 'is_after':
    return Object.assign(base, {
                 properties: {
                   [field]: {
	             type: 'string',
		     format: 'date-time',
		     const: normalizedValue,
		  },
		},
		required: [field]
    }
    default:
      return base;
}
```

Reply:
I know this was pre-existing, but since we're refactoring and it appears the simplification can be made
I also assume we can make this simplification can be made in all DataTypes

Reply:
I would avoid doing this, since it is only applicable to some DataTypes, others have multiple rules within the if or different rules, which does not allow us to use the switch case and would create a difference between files. I find it easier to read if they all stay consistent

Reply:
But in the case of the DataTypes with multiple rules, we could have multiple different cases in the switch instead of stacking them as above @JSReds 

It stays consistent in the sense that we're using `switch` for it everywhere and we also get the reduce LOC. Honestly I don't think have multiple identical if statements everywhere is clean. If I see separate if statements, I start hunting for the different between them.

Reply:
Looking more I see that date-time is a special case that happens to have all operators returning the same thing, but even in that case I would still at least dump all the conditions into 1 if statement rather than having 3 identical if statements

Reply:
> But in the case of the DataTypes with multiple rules, we could have multiple different cases in the switch instead of stacking them as above

how would you use a switch case in this case ?  

```
	if (
		operatorSlug === 'not_contains' &&
		!!schema.items &&
		typeof schema.items === 'object' &&
		'type' in schema.items &&
		!('properties' in schema.items)
	) {
```

Also cases like :

```
		operatorSlug === 'key_matches_re' ||
		operatorSlug === 'value_matches_re'

```

in the case of the object we would add many lines of code, and from the benefits that a switch case gives me it doesn't seem worth it. That said, it is my opinion, I can change the ones that can be changed if we all agree.

cons:

- array without switch case (inconsistent)
- object doubles the lines of code of the `createFilter` function (worth it ?)

Reply:
For the first case you mention, a switch would indeed be inappropriate
Wouldn't the second case just be this?:
```
switch(operatorSlug) {
  case 'key_matches_re':
  case 'value_matches_re':
    ...
}
```

Anyway I hear your point about the inconsistency of using a switch in some files but not in others (per the first case)
Do you feel like having 1 if statement like this would be inappropriate?:
```
if (['is', 'is_before', 'is_after'].includes(operatorSlug)) {
  ...
}
```

That manages to avoid the inconsistency of a switch in just 1 file while reducing LOC per a repetitive case

Reply:
I got lost here on this last piece of information. I probably don't remember the case any more, could you please remind me how this could reduce the lines of code ?
The 3 cases `['is', 'is_before', 'is_after']` are different. We could unify them using ternaries, but since I'm not 100% sure we listed all cases and that all properties in each case are the right one, I would keep them separated for now.

Reply:
```
if (operatorSlug === 'is') {

      
      


      
            
      
  
        
        

        
      

          

  
    
        
        
  
      
          
          
    

      
    
    
        
                👍
        
        
                👎
        
        
                😄
        
        
                🎉
        
        
                😕
        
        
                ❤️
        
        
                🚀
        
        
                👀
        
    

  

      

    
            
    


      
          
            
              
  
  

            
            
              
    

              Sorry, something went wrong.
            
            
             Quote reply
            
          
      
    
  

  
    

      
    Member


    

  

  


    

    
        myarmolinsky
  

    

    

    


      on Aug 1

    
  •

  
    
      
        
          edited
          
        
        
    

      
    
    
      
        
  
  

      
    
  


  


        
      
  
  
  
  
    
      
    

    

    There was a problem hiding this comment.


  

 
  
  
    
      
    

    

    Choose a reason for hiding this comment

    
      The reason will be displayed to describe this comment to others. Learn more.
    

    
      
        
        Choose a reason
        
        Spam
Abuse
Off Topic
Outdated
Duplicate
Resolved
      
            Hide comment


  


  
    
      Looks like we can just do the following, no?:
switch(operatorSlug) {
  case 'is':
  case 'is_before':
  case 'is_after':
    return Object.assign(base, {
                 properties: {
                   [field]: {
	             type: 'string',
		     format: 'date-time',
		     const: normalizedValue,
		  },
		},
		required: [field]
    }
  default:
    return base;
}

    
      
    

      
    

    
  
    
  
            



  
    
    Refresh and try again.
  




    

        
      
  
    
  
  

  
  
    
    

    Sorry, something went wrong.
  



        
      
  
    
    
      
        All reactions

            
          

    


    



  
        
  
      
          
          
    

      
    
    
        
                👍
        
        
                👎
        
        
                😄
        
        
                🎉
        
        
                😕
        
        
                ❤️
        
        
                🚀
        
        
                👀
        
    

  

      

    
            
    


      
          
            
              
  
  

            
            
              
    

              Sorry, something went wrong.
            
            
             Quote reply
            
          
      
    
  

  
    

      
    Member


    

  

  


    

    
        myarmolinsky
  

    

    

    


      on Aug 1

    
  


        
      
  
  
  
  
    
      
    

    

    There was a problem hiding this comment.


  

 
  
  
    
      
    

    

    Choose a reason for hiding this comment

    
      The reason will be displayed to describe this comment to others. Learn more.
    

    
      
        
        Choose a reason
        
        Spam
Abuse
Off Topic
Outdated
Duplicate
Resolved
      
            Hide comment


  


  
    
      I know this was pre-existing, but since we're refactoring and it appears the simplification can be made
I also assume we can make this simplification can be made in all DataTypes
    
  
            



  
    
    Refresh and try again.
  




    

        
      
  
    
  
  

  
  
    
    

    Sorry, something went wrong.
  



        
      
  
    
    
      
        All reactions

            
          

    


    



  
        
  
      
          
          
    

      
    
    
        
                👍
        
        
                👎
        
        
                😄
        
        
                🎉
        
        
                😕
        
        
                ❤️
        
        
                🚀
        
        
                👀
        
    

  

      

    
            
    


      
          
            
              
  
  

            
            
              
    

              Sorry, something went wrong.
            
            
             Quote reply
            
          
      
    
  

  
    

      
    Member


    Author

  

  


    

    
        JSReds
  

    

    

    


      on Aug 2

    
  


        
      
  
  
  
  
    
      
    

    

    There was a problem hiding this comment.


  

 
  
  
    
      
    

    

    Choose a reason for hiding this comment

    
      The reason will be displayed to describe this comment to others. Learn more.
    

    
      
        
        Choose a reason
        
        Spam
Abuse
Off Topic
Outdated
Duplicate
Resolved
      
            Hide comment


  


  
    
      I would avoid doing this, since it is only applicable to some DataTypes, others have multiple rules within the if or different rules, which does not allow us to use the switch case and would create a difference between files. I find it easier to read if they all stay consistent
    
  
            



  
    
    Refresh and try again.
  




    

        
      
  
    
  
  

  
  
    
    

    Sorry, something went wrong.
  



        
      
  
    
    
      
        All reactions

            
          

    


    



  
        
  
      
          
          
    

      
    
    
        
                👍
        
        
                👎
        
        
                😄
        
        
                🎉
        
        
                😕
        
        
                ❤️
        
        
                🚀
        
        
                👀
        
    

  

      

    
            
    


      
          
            
              
  
  

            
            
              
    

              Sorry, something went wrong.
            
            
             Quote reply
            
          
      
    
  

  
    

      
    Member


    

  

  


    

    
        myarmolinsky
  

    

    

    


      on Aug 2

    
  •

  
    
      
        
          edited
          
        
        
    

      
    
    
      
        
  
  

      
    
  


  


        
      
  
  
  
  
    
      
    

    

    There was a problem hiding this comment.


  

 
  
  
    
      
    

    

    Choose a reason for hiding this comment

    
      The reason will be displayed to describe this comment to others. Learn more.
    

    
      
        
        Choose a reason
        
        Spam
Abuse
Off Topic
Outdated
Duplicate
Resolved
      
            Hide comment


  


  
    
      But in the case of the DataTypes with multiple rules, we could have multiple different cases in the switch instead of stacking them as above @JSReds
It stays consistent in the sense that we're using switch for it everywhere and we also get the reduced LOC. Honestly I don't think having multiple identical if statements everywhere is clean. If I see separate if statements, I start hunting for the difference between them.
    
  
            



  
    
    Refresh and try again.
  




    

        
      
  
    
  
  

  
  
    
    

    Sorry, something went wrong.
  



        
      
  
    
    
      
        All reactions

            
          

    


    



  
        
  
      
          
          
    

      
    
    
        
                👍
        
        
                👎
        
        
                😄
        
        
                🎉
        
        
                😕
        
        
                ❤️
        
        
                🚀
        
        
                👀
        
    

  

      

    
            
    


      
          
            
              
  
  

            
            
              
    

              Sorry, something went wrong.
            
            
             Quote reply
            
          
      
    
  

  
    

      
    Member


    

  

  


    

    
        myarmolinsky
  

    

    

    


      on Aug 2

    
  


        
      
  
  
  
  
    
      
    

    

    There was a problem hiding this comment.


  

 
  
  
    
      
    

    

    Choose a reason for hiding this comment

    
      The reason will be displayed to describe this comment to others. Learn more.
    

    
      
        
        Choose a reason
        
        Spam
Abuse
Off Topic
Outdated
Duplicate
Resolved
      
            Hide comment


  


  
    
      Looking more I see that date-time is a special case that happens to have all operators returning the same thing, but even in that case I would still at least dump all the conditions into 1 if statement rather than having 3 identical if statements
    
  
            



  
    
    Refresh and try again.
  




    

        
      
  
    
  
  

  
  
    
    

    Sorry, something went wrong.
  



        
      
  
    
    
      
        All reactions

            
          

    


    





  
        
  
      
          
          
    

      
    
    
        
                👍
        
        
                👎
        
        
                😄
        
        
                🎉
        
        
                😕
        
        
                ❤️
        
        
                🚀
        
        
                👀
        
    

  

      

    
            
    


      
          
            
              
  
  

            
            
              
    

              Sorry, something went wrong.
            
            
             Quote reply
            
          
      
    
  

  
    

      
    Member


    Author

  

  


    

    
        JSReds
  

    

    

    


      29 days ago

    
  


        
      
  
  
  
  
    
      
    

    

    There was a problem hiding this comment.


  

 
  
  
    
      
    

    

    Choose a reason for hiding this comment

    
      The reason will be displayed to describe this comment to others. Learn more.
    

    
      
        
        Choose a reason
        
        Spam
Abuse
Off Topic
Outdated
Duplicate
Resolved
      
            Hide comment


  


  
    
      
But in the case of the DataTypes with multiple rules, we could have multiple different cases in the switch instead of stacking them as above

how would you use a switch case in this case ?
	if (
		operatorSlug === 'not_contains' &&
		!!schema.items &&
		typeof schema.items === 'object' &&
		'type' in schema.items &&
		!('properties' in schema.items)
	) {

    
      
    

      
    

    
  
Also cases like :
		operatorSlug === 'key_matches_re' ||
		operatorSlug === 'value_matches_re'


    
      
    

      
    

    
  
in the case of the object we would add many lines of code, and from the benefits that a switch case gives me it doesn't seem worth it. That said, it is my opinion, I can change the ones that can be changed if we all agree.
cons:

array without switch case (inconsistent)
object doubles the lines of code of the createFilter function (worth it ?)

    
  
            



  
    
    Refresh and try again.
  




    

        
      
  
    
  
  

  
  
    
    

    Sorry, something went wrong.
  



        
      
  
    
    
      
        All reactions

            
          

    


    



  
        
  
      
          
          
    

      
    
    
        
                👍
        
        
                👎
        
        
                😄
        
        
                🎉
        
        
                😕
        
        
                ❤️
        
        
                🚀
        
        
                👀
        
    

  

      

    
            
    


      
          
            
              
  
  

            
            
              
    

              Sorry, something went wrong.
            
            
             Quote reply
            
          
      
    
  

  
    

      
    Member


    

  

  


    

    
        myarmolinsky
  

    

    

    


      29 days ago

    
  


        
      
  
  
  
  
    
      
    

    

    There was a problem hiding this comment.


  

 
  
  
    
      
    

    

    Choose a reason for hiding this comment

    
      The reason will be displayed to describe this comment to others. Learn more.
    

    
      
        
        Choose a reason
        
        Spam
Abuse
Off Topic
Outdated
Duplicate
Resolved
      
            Hide comment


  


  
    
      For the first case you mention, a switch would indeed be inappropriate
Wouldn't the second case just be this?:
switch(operatorSlug) {
  case 'key_matches_re':
  case 'value_matches_re':
    ...
}

    
      
    

      
    

    
  
Anyway I hear your point about the inconsistency of using a switch in some files but not in others (per the first case)
Do you feel like having 1 if statement like this would be inappropriate?:
if (['is', 'is_before', 'is_after'].includes(operatorSlug)) {
  ...
}

    
      
    

      
    

    
  
That manages to avoid the inconsistency of a switch in just 1 file while reducing LOC per a repetitive case
    
  
            



  
    
    Refresh and try again.
  




    

        
      
  
    
  
  

  
  
    
    

    Sorry, something went wrong.
  



        
      
  
    
    
      
        All reactions

            
          

    


    



  
        
  
      
          
          
    

      
    
    
        
                👍
        
        
                👎
        
        
                😄
        
        
                🎉
        
        
                😕
        
        
                ❤️
        
        
                🚀
        
        
                👀
        
    

  

      

    
            
    


      
          
            
              
  
  

            
            
              
    

              Sorry, something went wrong.
            
            
             Quote reply
            
          
      
    
  

  
    

      
    Member


    Author

  

  


    

    
        JSReds
  

    

    

    


      21 hours ago

    
  


        
      
  
  
  
  
    
      
    

    

    There was a problem hiding this comment.


  

 
  
  
    
      
    

    

    Choose a reason for hiding this comment

    
      The reason will be displayed to describe this comment to others. Learn more.
    

    
      
        
        Choose a reason
        
        Spam
Abuse
Off Topic
Outdated
Duplicate
Resolved
      
            Hide comment


  


  
    
      I got lost here on this last piece of information. I probably don't remember the case any more, could you please remind me how this could reduce the lines of code ?
The 3 cases ['is', 'is_before', 'is_after'] are different. We could unify them using ternaries, but since I'm not 100% sure we listed all cases and that all properties in each case are the right one, I would keep them separated for now.
    
  
            



  
    
    Refresh and try again.
  




    

        
      
  
    
  
  

  
  
    
    

    Sorry, something went wrong.
  



        
      
  
    
    
      
        All reactions

            
          

    


    




    

      
        
          
            
              
                
              
              
                
                  Reply...
                
              
            
          

          
            
              
              
              
              
              
              
              

              
  
  
      
        
          Write
        
        
          Preview
        
      
    

        
        
  
      
    

  Add a suggestion, <Ctrl+g>


      


  
    
      
    

    
    Add heading text

    
      
    

    
    Add bold text, <Ctrl+b>

    
      
    

    
    Add italic text, <Ctrl+i>
  

  
    
      
    

    
    Add a quote, <Ctrl+Shift+.>

    
      
    

    
    Add code, <Ctrl+e>



    
      
    

    
    Add a link, <Ctrl+k>
  

  
    
      
    

    
    Add a bulleted list, <Ctrl+Shift+8>

    
      
    

    
    Add a numbered list, <Ctrl+Shift+7>

    
      
    

    
    Add a task list, <Ctrl+Shift+l>
  

        

          
            
    

          
          Directly mention a user or team


          
            
    

          
          Reference an issue, pull request, or discussion

          
  
    
    

  

  
    
      Select a reply
      ctrl .
    

    
      
  
  

    

  

Add saved reply


      



  

  
    There was an error creating your PullRequest.
  

  
    
        

      
          
          

      
          
    
    
    
      
        Attach files by dragging & dropping, selecting or pasting them.
      
      
        
  
  

          
            Uploading your files…
          
      
      
        We don’t support that file type.
        
          Try again with a
          GIF, JPEG, JPG, MOV, MP4, PNG, SVG, WEBM, CSV, DOCX, FODG, FODP, FODS, FODT, GZ, LOG, MD, ODF, ODG, ODP, ODS, ODT, PDF, PPTX, TGZ, TXT, XLS, XLSX or ZIP.
        
      
      
        Attaching documents requires write permission to this repository.
        
          Try again with a GIF, JPEG, JPG, MOV, MP4, PNG, SVG, WEBM, CSV, DOCX, FODG, FODP, FODS, FODT, GZ, LOG, MD, ODF, ODG, ODP, ODS, ODT, PDF, PPTX, TGZ, TXT, XLS, XLSX or ZIP.
        
      
      
        We don’t support that file type.
        
          Try again with a GIF, JPEG, JPG, MOV, MP4, PNG, SVG, WEBM, CSV, DOCX, FODG, FODP, FODS, FODT, GZ, LOG, MD, ODF, ODG, ODP, ODS, ODT, PDF, PPTX, TGZ, TXT, XLS, XLSX or ZIP.
        
      
      
      
      
        This file is empty.
        
          Try again with a file that’s not empty.
        
      
      
        This file is hidden.
        
          Try again with another file.
        
      
      
        Something went really wrong, and we can’t process that file.
        
          Try again.
        
      
    
    
      
    

    
    Styling with Markdown is supported
  

    

  
    
    
    
    
    
    
    
    
    
    
  
    Nothing to preview
  


  


  



              
                
                          Add review comment
                      Start a review

                          Add single comment

                

                      Cancel

              
          
        
      
  

    
      
            Resolve conversation






    

      
  
		return Object.assign(base, {
			properties: {
				[field]: {
					type: 'string',
					format: 'date-time',
					const: normalizedValue,
				},
			},
			required: [field],
		});
	}

	if (operatorSlug === 'is_before') {
		return Object.assign(base, {
			properties: {
				[field]: {
					type: 'string',
					format: 'date-time',
					formatMaximum: normalizedValue,
				},
			},
			required: [field],
		});
	}

	if (operatorSlug === 'is_after') {
		return Object.assign(base, {
			properties: {
				[field]: {
					type: 'string',
					format: 'date-time',
					formatMinimum: normalizedValue,
				},
			},
			required: [field],
		});
	}
```
It looks like they have the exact same content in the if statements? @JSReds

Reply:
Argh I see that they're different know 🤦 
They looked the same but they have `const` vs `formatMaximum` vs `formatMinimum` in them 🤦 🤦 🤦 
*sigh* apologies for this huge mess 😞 , the code looks good